Wipe All Of The Things!

kaapstorm opened the pull request:
[SC-514](https://dimagi-dev.atlassian.net/browse/SC-514): "Come up with data wipe strategies for each persistent service"

See also [related CommCare Cloud PR](https://github.com/dimagi/commcare-cloud/pull/3770) 

##### SUMMARY
Adds three Django management commands for wiping persistent data:

* **list_kafka_topics** used by Ansible to wipe all Kafka topics
* **wipe_blobdb** wipes all blobs
* **wipe_es** drops all Elasticsearch indices.

kaapstorm committed:
Command to list all Kafka topics

kaapstorm committed:
Command to wipe BlobDB

kaapstorm committed:
Command to wipe Elasticsearch

kaapstorm commented:
Particularly curious about your opinion on `wipe_blobdb.py` ... you'll see why :smirk:

I am mulling over some starting content for a "HQ Guidelines" doc about what I've done there.

millerdev commented on corehq/blobs/management/commands/wipe_blobdb.py:
```diff
@@ -0,0 +1,70 @@
+
+async def wipe_shard(dbname, commit=False):
+    bytes_deleted = 0
+    metas = BlobMeta.objects.using(dbname).all()
```
This could be a really huge result. It will probably be better to use a pagination strategy. Ideally it would do the SUM in a single query along with the DELETE in Postgres. Something like

```sql
WITH deleted AS (
    DELETE FROM blobs_blobmeta
    WHERE id IN (
        SELECT id
        FROM blobs_blobmeta
        ORDER BY id
        LIMIT 1000
    )
    RETURNING *
)
SELECT SUM(content_length) AS bytes_deleted
FROM deleted;
```

kaapstorm replied:
Isn't it paginated? I thought `all()` is lazy, and `chunked()` is only fetching 1000 rows at a time. ... Or am I misunderstanding something?

millerdev replied:
I don't think `all()` is lazy, or at least it loads all of the results immediately when the first one is accessed.

kaapstorm replied:
Ah! You are right. [QuerySets are evaluated](https://docs.djangoproject.com/en/1.11/ref/models/querysets/#when-querysets-are-evaluated) when they are iterated, or when they are sliced (among other things).

Django's `Paginator` class [uses slicing](https://github.com/django/django/blob/master/django/core/paginator.py#L78) to evaluate them late ... but should not be used on [large QuerySets](https://docs.djangoproject.com/en/3.0/ref/paginator/#django.core.paginator.Paginator.object_list), so I won't use it here.

I'll change this to use slicing to iterate chunks.

millerdev replied:
> I'll change this to use slicing to iterate chunks.

I think even Django pagination logic will not work well for very large result sets because it uses limit with *offset* (`self.object_list[bottom:top]`), which does not perform well for large offsets.

[Posgres docs](https://www.postgresql.org/docs/8.3/queries-limit.html) say:
> The rows skipped by an `OFFSET` clause still have to be computed inside the server; therefore a large `OFFSET` might be inefficient.

The same issue is referenced in this article https://www.citusdata.com/blog/2016/03/30/five-ways-to-paginate/

kaapstorm replied:
The Postgres docs explain it better (for my brain at least) than the Django docs. :facepalm: I've just done the same thing, only shorter.

[Disqus solved this](http://cra.mr/2011/03/08/building-cursors-for-the-disqus-api) by using the primary key instead of offset. `BlobMeta`'s primary key is a normal auto-incrementing integer field called `id`, so a paginator can use that.

esoergel reviewed (approved):
looks good to me, aside from Daniel's comments on the blobdb approach

kaapstorm committed:
Paginate using slicing

... because "slicing an unevaluated QuerySet returns another
unevaluated QuerySet"

https://docs.djangoproject.com/en/1.11/ref/models/querysets/#when-querysets-are-evaluated

kaapstorm committed:
Paginate using `id` instead of offset
